### PR TITLE
Add hierarchical module names to torchFX graph.node #87659

### LINKED ISSUE #87659: Preserve module hierarchy information in torch.FX graph.nodes
### 🚀 The feature, motivation and pitch

The module hierarchy in PyTorch networks is saved in nn.module.named_modules().  This function provides a unique name that maps to each module object in the network.  However, when torch.fx.symbolic_trace() is run on a network, the module hierarchy is lost in the resulting torch.fx.GraphModule.    The unique module names effectively describe the network architecture.  Here's an example name from Hugging Face stable diffusion
**text_model.encoder.layers.0.self_attn.mul** . torchFX would convert this name to just **mul_1** .   From the HF example, just by looking at the module name we can tell the op is part of the encoder from the 0th layer in the self_attn block.  Whereas with the name **mul_1** we can't tell what part of the network this operator belongs to.  

In torch.fx the network graph is stored in the GraphModule.graph.nodes class.  However, the nodes are just a flat listing of all the operators in the network, they don't include any references to the modules which defined the nodes.  The PyTorch ONNX exporter was recently updated to include the module information in the operator unique names.  This feature request is proposing to do the same for torch.FX.

### Alternatives

Live with the current default node names and refer directly to the network source code to try and understand the network architecture.

### Additional context

Here's an example of what the unique operator (graph.node) names should look like.  Just by viewing the graph or the listing from graph.graph you can understand the network architecture which is much easier to work with than inspecting the network source code.

image

cc @ezyang @SherlockNoMad @soumith @EikanWang @jgong5 @wenzhe-nrv

### PR DESCRIPTION
**Pull request by alexmsettle**
Fixes #87659

Pass down the module hierarchy from module.named_modules() to the name field of graph.node.
This makes it so the name of each node contains descriptive information about the network architecture. 

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

### CODE REVIEW

**Comment by ezyang** [comment]
@SherlockNoMad bounce this back to me if you're not the right reviewer

File: torch/fx/_symbolic_trace.py [hunk 1]
```diff
@@ -430,9 +489,29 @@ def call_module(
...
+        kwargs: Dict[str, Argument],
+        name: Optional[str] = None,
+        type_expr: Optional[Any] = None
+    ) -> Node:
+        node = super().create_node(kind, target, args, kwargs, name, type_expr)
+        self.node_name_to_scope[node.name] = (
+            self.scope.module_path,
+            self.scope.module_type,
+        )
+        node.meta['nn_module_stack'] = {self.scope.module_path : self.scope.module_type}
```
Comment (SherlockNoMad) [inline torch/fx/_symbolic_trace.py:513]: This is different from what is store in nn_module_stack from the dynamo trace... 
Ideally, we want to make it consistent... 

node.meta['nn_module_stack'] is currently an OrderedDictionary that store the entire stack of modules. 

@jerryzh168, do you have a preference which structure it should take?

Comment (jerryzh168) [reply torch/fx/_symbolic_trace.py:513]: I think today we actually only use a single level of scope (modue_fqn, module_type) in quantization (e.g. ("sub.conv", nn.Conv2d)) (actually the scope context code in this PR is the same as the ones defined in quantization tracer: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/tracer.py#L11), but I think theoretically we should be storing a stack of (module_fqn, module_type) 
e.g. 
for a module like following:
class Sub(...):
    def __init__(self, ...):
          self.conv = nn.Conv2d(...)

class M(...):
    def __init__(self, ...):
          self.sub = Sub()

     def forward(self, x):
           return self.sub(x)

for the conv aten op, the scope should be something like
[("sub.conv", nn.Conv2d), ("sub", Sub)] since to be precise, the conv aten node is in the scope of the Conv module and Sub module. We can get module fqn for sub just by looking at "sub.conv", but we can't get the module type of "Sub" easily if we just have a single entry of ("sub.conv", nn.Conv2d)

Comment (alexmsettle) [reply torch/fx/_symbolic_trace.py:513]: FYI - I actually just moved the Scope related code from quantization tracer into the base class Tracer so everyone has access to the module scope. 

Doesn't the module_fqn already account for the hierarchy?  Eg when the aten::conv is actually called the module_fqn would be sub.conv and the op would be aten::conv2d.  Ok now I see the issue, if you want to preserve the module type at each level of the hierarchy then you do need a listing of each level of hierarchy.

File: torch/fx/proxy.py [hunk 2]
```diff
@@ -13,6 +13,64 @@
...
+
+    def __init__(
+        self, scope: Scope, current_module: torch.nn.Module, current_module_path: str
+    ):
+        super().__init__()
+        self.prev_module_type = scope.module_type
+        self.prev_module_path = scope.module_path
+        self.scope = scope
+        self.scope.module_path = current_module_path
+        self.scope.module_type = type(current_module)
```
Comment (jerryzh168) [inline torch/fx/proxy.py:63]: nit: I feel these can probably be simplified by just storing previous scope object and current scope object

Comment (alexmsettle) [reply torch/fx/proxy.py]: I just fixed that.  ^^^

Working on capturing the scope stack now.

File: torch/fx/_symbolic_trace.py (hunk 1 above)
Comment (alexmsettle) [reply torch/fx/_symbolic_trace.py]: @SherlockNoMad  I just added an ordered dict for module stack.  Now the code writes an ordered dict to node.meta['nn_module_stack']

For the example code above w/ Module and Sub module I get this output  `Node sub_mod_conv_0 - mod {'sub_mod': "<class '__main__.SubMod'>", 'sub_mod.conv_0': "<class 'torch.nn.modules.conv.Conv2d'>"}`

**Comment by alexmsettle** [comment]
ping @jerryzh168 @SherlockNoMad - I think I addressed all the issues raised in the code review.   Should I add a unit test to protect the module names?  

I had to move much of my changes to the TracerBase class because of backward compatibility testing issues.

File: torch/fx/proxy.py [hunk 3]
```diff
@@ -11,7 +12,68 @@
...
+
+@compatibility(is_backward_compatible=False)
+class ScopeContextManager(object):
+    """ A context manager to track the Scope of Node during symbolic tracing.
+    When entering a forward function of a Module, we'll update the scope information of
+    the current module, and when we exit, we'll restore the previous scope information.
+    """
+
+    def __init__(
+        self, scope: Scope, current_module: torch.nn.Module, current_module_path: str
```
Comment (jerryzh168) [inline torch/fx/proxy.py:59]: nit: just to make this clearer, should we use
```
def __init__(self, prev_scope: Scope, current_scope: Scope):
```?

File: torch/fx/proxy.py [hunk 4]
```diff
@@ -11,7 +12,68 @@
...
+    ):
+        super().__init__()
+        self._prev_scope = Scope("", None)
+        self._prev_scope.module_type = scope.module_type
+        self._prev_scope.module_path = scope.module_path
+        self.scope = scope
+        self.scope.module_path = current_module_path
+        self.scope.module_type = type(current_module)
```
Comment (jerryzh168) [inline torch/fx/proxy.py:67]: this part feels confusing, we can probably cleanup this after we change the signature as well

File: torch/fx/proxy.py [hunk 5]
```diff
@@ -11,7 +12,68 @@
...
+
+    def __enter__(self):
+        return self.scope
+
+    def __exit__(self, *args):
+        self.scope.module_path = self._prev_scope.module_path
+        self.scope.module_type = self._prev_scope.module_type
```
Comment (jerryzh168) [inline torch/fx/proxy.py:74]: feels like this could just be `self.scope = self._prev_scope`?

File: torch/fx/proxy.py (hunk 4 above)
Comment (alexmsettle) [reply torch/fx/proxy.py:67]: Fixed this by changing the signature from previous comment.

File: torch/fx/proxy.py (hunk 5 above)
Comment (alexmsettle) [reply torch/fx/proxy.py:74]: Fixed this too.

**Comment by alexmsettle** [comment]
Thanks for reviewing my changes @jerryzh168 .  I was on vacation last week, so only just got to it today.

I also just added a test case to protect the node.meta['nn_module_stack']

**Comment by alexmsettle** [comment]
Hey @jerryzh168 @SherlockNoMad  I think these changes are ready to go.  Could you please review?

File: torch/fx/proxy.py [hunk 6]
```diff
@@ -11,7 +12,71 @@
...
+    """ A context manager to track the Scope of Node during symbolic tracing.
+    When entering a forward function of a Module, we'll update the scope information of
+    the current module, and when we exit, we'll restore the previous scope information.
+    """
+
+    def __init__(
+        self, scope: Scope,
+        current_module: torch.nn.Module,
+        current_module_path: str
+    ):
```
Comment (jerryzh168) [inline torch/fx/proxy.py:62]: nit: I think it might be clearer if we just pass around the Scope object?

something like the following:

```
def __init__(
        self, 
        prev_scope: Scope,
        current_scope: Scope,
    ):
```

File: torch/fx/proxy.py [hunk 7]
```diff
@@ -43,7 +117,14 @@ def create_node(self, kind : str, target : Target,
         if kind == 'call_function' and self.check_mutable_operations:
             check_for_mutable_operation(target, args, kwargs)
 
-        return self.graph.create_node(kind, target, args, kwargs, name, type_expr)
+        node = self.graph.create_node(kind, target, args, kwargs, name, type_expr)
+        self.node_name_to_scope[node.name] = (
+            self.scope.module_path,
+            self.scope.module_type,
+        )
```
Comment (jerryzh168) [inline torch/fx/proxy.py:124]: maybe we can add a TODO to remove this, it now becomes redundant since we have `node.meta["nn_module_stack"]`

File: torch/fx/_symbolic_trace.py [hunk 8]
```diff
@@ -250,6 +249,13 @@ def __init__(
         self.param_shapes_constant = param_shapes_constant
 
         self.submodule_paths: Optional[Dict[torch.nn.Module, str]] = None
+        self.root_module_name: str = ""
+        # Maps the containing module's name to the operator name
+        self.scope = Scope("", None)
+        # Records the module call stack
+        self.module_stack = {}
```
Comment (jerryzh168) [inline torch/fx/_symbolic_trace.py:256]: should this be an OrderedDict?

Comment (jerryzh168) [reply torch/fx/_symbolic_trace.py:256]: also nit: maybe rename this to `nn_module_stack` to make it clearer

File: torch/fx/_symbolic_trace.py [hunk 9]
```diff
@@ -430,9 +436,17 @@ def call_module(
             value was returned from the ``Module`` invocation.
         """
         module_qualified_name = self.path_of_module(m)
-        if not self.is_leaf_module(m, module_qualified_name):
-            return forward(*args, **kwargs)
-        return self.create_proxy("call_module", module_qualified_name, args, kwargs)
+        with ScopeContextManager(self.scope, m, module_qualified_name) as _scope:
+            # module_stack is an ordered dict so writing then deleting the
+            # entry is equivalent to push/pop on a list
+            self.module_stack[_scope.module_path] = str(_scope.module_type)
```
Comment (jerryzh168) [inline torch/fx/_symbolic_trace.py:442]: actually, it feels like we can probably maintain this thing in a scope as well

File: test/test_fx.py [hunk 10]
```diff
@@ -1679,6 +1679,36 @@ def forward(self, x):
...
+        gm = torch.fx.symbolic_trace(m)
+
+        mod_stack = {}
+        expected_stack = [{'sub_mod': str(type(m.sub_mod))},
+                          {'sub_mod.conv_mod': str(type(m.sub_mod.conv_mod))}]
+        for node in gm.graph.nodes:
+            mod_stack = node.meta.get('nn_module_stack', {})
+            if mod_stack:
+                break
+        stack_list = [{item: mod_stack[item]} for item in mod_stack]
```
[the remark this answers is not in the document]
Comment (alexmsettle) [reply test/test_fx.py]: Done

File: torch/fx/proxy.py (hunk 6 above)
Comment (alexmsettle) [reply torch/fx/proxy.py]: Done

File: torch/fx/_symbolic_trace.py (hunk 8 above)
Comment (alexmsettle) [reply torch/fx/_symbolic_trace.py]: Added typing.OrderedDict to initialization

Done

File: torch/fx/_symbolic_trace.py (hunk 9 above)
Comment (alexmsettle) [reply torch/fx/_symbolic_trace.py:443]: The self.module_stack is the data structure that populates node.meta['nn_module_stack'].  It is supposed to be an OrderedDict to be consistent with torch dynamo logic that @SherlockNoMad implemented.

**Comment by alexmsettle** [comment]
Hey @jerryzh168  - I made the changes you requested and just synced with the main pytorch branch.  All tests are passing. Can you please review?

File: torch/fx/proxy.py [hunk 11]
```diff
@@ -1,17 +1,83 @@
...
+        self,
+        prev_scope: Scope,
+        current_scope: Scope,
+    ):
+        super().__init__()
+        # Make a copy of scope to restore on context block exit
+        self._prev_scope = copy.copy(prev_scope)
+        # Modify the prev_scope that was passed in by reference
+        prev_scope.module_path = current_scope.module_path
+        prev_scope.module_type = current_scope.module_type
```
Comment (jerryzh168) [inline torch/fx/proxy.py:69]: why do we have to do this?

Comment (jerryzh168) [reply torch/fx/proxy.py:69]: OK I looked again, it looks like `prev_scope` is actually a `scope` variable that keeps track of current scope, maybe we should update the name to reflect this better, e.g. update `prev_scope` to just `scope`

Comment (jerryzh168) [reply torch/fx/proxy.py:69]: I feel something like the following might be clearer (pretty similar to original code)
```
def __init__(
        self,
        scope: Scope,
        current_scope: Scope,
    ):
    # keep a copy of previous scope so that we can restore the scope
    self._prev_scope = copy.copy(scope)
    # update scope to current scope
    scope.module_path = current_scope.module_path
    scope.module_type = current_scope.module_type
    # holding a reference so that we can restore it
    self.scope = scope

...

def __exit__(self, *args):
        # restore the scope to previous scope
        self.scope.module_path = self._prev_scope.module_path
        self.scope.module_type = self._prev_scope.module_type
        return
```

Comment (alexmsettle) [reply torch/fx/proxy.py:69]: You beat me to it.  Your comments above make sense.  Using the name prev_scope made the code confusing.

**Comment by alexmsettle** [comment]
Ok @jerryzh168 - I made the above changes and all tests are passing.  Is this change set ready for sign off?

**Comment by alexmsettle** [comment]
Ping @jerryzh168  - I just synced from the main upstream branch and the tests pass.  Could you pls take another look so we can get this merged?

**Review by jerryzh168 — APPROVED** [review]
LGTM, thanks for addressing all the comments!

File: torch/fx/_symbolic_trace.py [hunk 12]
```diff
@@ -250,6 +250,13 @@ def __init__(
         self.param_shapes_constant = param_shapes_constant
 
         self.submodule_paths: Optional[Dict[torch.nn.Module, str]] = None
+        self.root_module_name: str = ""
+        # Maps the containing module's name to the operator name
+        self.scope = Scope("", None)
+        # Records the module call stack
+        self.module_stack = collections.OrderedDict()
+        # Mapping of node name to module scope
+        self.node_name_to_scope: Dict[str, Tuple[str, type]] = {}
```
Comment (jerryzh168) [inline torch/fx/_symbolic_trace.py:259]: another thing we could do is just to attach the scope to node itself, but this will involve quantization code as well and we can do it in a separate PR

Comment (alexmsettle) [reply torch/fx/_symbolic_trace.py:259]: ok that sounds good.

**Comment by alexmsettle** [comment]
> LGTM, thanks for addressing all the comments!

Thanks for all the feedback and reviews @jerryzh168
